New lint: `decimal_bit_mask`

Artur-Sulej opened the pull request:
changelog: [`decimal_bit_mask`]: add a new lint that detects the use of decimal literals as bit masks in bitwise operations. Using decimal literals for bit masks can obscure the intended bit pattern and reduce code readability. This lint encourages the use of binary (`0b...`) or hexadecimal (`0x...`) notation to make bit patterns explicit and easier to understand at a glance. 

---
Example:
```rust
let masked = x & 15; // Bad: decimal literal as bit mask
let masked = x & 0b1111; // Good: bit pattern is explicit
```

Artur-Sulej commented:
Although I annotated my test with `//~ decimal_bit_mask`, Clippy still reports the lint as an unmatched diagnostic and the test fails.  
Any idea what I might be missing?

ada4a commented:
@rustbot note remove Feature-freeze

Artur-Sulej commented on clippy_lints/src/decimal_bit_mask.rs:
```diff
@@ -0,0 +1,106 @@
+}
+
+fn is_not_decimal_number(cx: &LateContext<'_>, span: Span) -> bool {
+    span.check_source_text(cx, |src| src.contains("0b") || src.contains("0x") || src.contains("0o"))
```
> Here, 0b1000 isn't recognized as being a binary literal, because the way you check for that is by looking at the span of the literal, but here the span actually contains the parentheses surrounding the literal. Because of that, you'll need to first trim starting paren(s), then possibly some whitespace, and only then performing the actual checks.

@ada4a I simplified this part (using `contains`) to handle parentheses, whitespaces and comments. It might miss some complex cases, but that’s an intentional trade-off. Such cases usually mean special constructs and skipping the lint seems more appropriate.

ada4a commented on clippy_lints/src/decimal_bit_mask.rs:
```diff
@@ -0,0 +1,106 @@
+        span,
+        "using decimal literal for bitwise operation",
+        None,
+        "use binary (0b...) or hex (0x...) or octal (0o...) notation for better readability",
```
```suggestion
        "use binary (0b...), hex (0x...), or octal (0o...) notation for better readability",
```

ada4a commented on clippy_lints/src/decimal_bit_mask.rs:
```diff
@@ -0,0 +1,106 @@
+            kind: ExprKind::Lit(lit),
+            ..
+        },
+    ) = &e.kind
```
very much a nit: imo, these big patterns are too verbose -- what do you think about something like this?
```suggestion
    if let ExprKind::AssignOp(op, _, expr) = &e.kind
        && matches!(op.node, AssignOpKind::BitAndAssign | AssignOpKind::BitOrAssign | AssignOpKind::BitXorAssign)
        && let ExprKind::Lit(lit) = expr.kind
```

ada4a commented on clippy_lints/src/decimal_bit_mask.rs:
```diff
@@ -0,0 +1,106 @@
+    #[clippy::version = "1.87.0"]
+    pub DECIMAL_BIT_MASK,
+    nursery,
+    "use binary, hex or octal literals for bitwise operations"
```
```suggestion
    "use binary, hex, or octal literals for bitwise operations"
```

ada4a commented on clippy_lints/src/decimal_bit_mask.rs:
````diff
@@ -0,0 +1,106 @@
+    /// let a = 0b1110 & 0b0110;
+    /// ```
+    #[clippy::version = "1.87.0"]
+    pub DECIMAL_BIT_MASK,
````
Forgot to say this earlier, but lint names should be in plural, see https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

ada4a reviewed (changes requested):
Great progress! Much smaller comments this time (apart from the move-to-`operators` thing, that might be a bit churny -- sorry about that^^)

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15215/0ac66461bb73d59ecec00e4f333a04838a6b5876..35d6cc09a3d5fbf71c7ad490897c2b7838ee82f0)*

Artur-Sulej commented on clippy_lints/src/operators/decimal_bitwise_operands.rs:
```diff
@@ -0,0 +1,65 @@
+
+use super::DECIMAL_BITWISE_OPERANDS;
+
+pub(super) fn check_binary<'tcx>(cx: &LateContext<'tcx>, op: BinOpKind, left: &'tcx Expr<'_>, right: &'tcx Expr<'_>) {
```
After moving this lint to the operators category (https://github.com/rust-lang/rust-clippy/pull/15215/commits/a6c0cfd6071a0968707d3221d24f60a9f95eb896), I decided to split the `check` function into `check_binary` and `check_assign`. This makes it more efficient, because it avoids redundant pattern matching that is already performed in `src/operators/mod.rs`.

ada4a commented on tests/ui/decimal_bitwise_operands.stderr:
```diff
@@ -0,0 +1,124 @@
+LL |     88 & 99;
+   |          ^^
+   |
+   = help: use binary (0b...), hex (0x...), or octal (0o...) notation for better readability
```
(Not sure about this)

Maybe it would make sense to lint expressions like this in one go, i.e. highlight both operands? Let me know you what you think; maybe a review could weigh in as well

ada4a commented on clippy_lints/src/operators/decimal_bitwise_operands.rs:
```diff
@@ -0,0 +1,65 @@
+    })
+}
+
+fn is_power_of_twoish(lit: &Spanned<LitKind>) -> bool {
```
very much a nit: you could pass `lit.node` at the callsite to avoid the need to take `Spanned` here (and thus import it)

llogiq commented on tests/ui/decimal_bitwise_operands.stderr:
```diff
@@ -0,0 +1,164 @@
+LL |     x | (/* comment */99);
+   |                       ^^
+   |
+   = help: use binary (0b1100011), hex (0x63), or octal (0o143) notation for better readability
```
Would it make sense to group 4 or at least 8 binary digits by underscore? Makes it easier to read and avoids having another lint triggering on applying this one.

llogiq reviewed (commented):
One question remains, otherwise this looks ok.

I'll start the Final Comments Period shortly.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15215/24e16f992c2dda1904fe8d4a94b391513832ce60..98f010db6f9824d692e2f3b54d78cb1778c63435)*

Artur-Sulej commented:
I addressed the comment and updated the lint message to group digits with underscores.

Additional changes:
- Reused the existing logic for detecting decimals (`NumericLiteral`), instead of custom parsing
- Extended handling of bitwise operations to cover unary and cast expressions

llogiq reviewed (approved):
I think the code is fine now, and will start the FCP when I get around to it (I recently only narrowly avoided being buried by a ceiling heating falling down and have still to deal with the aftermath).

Can you please squash your commits?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15215/24e16f992c2dda1904fe8d4a94b391513832ce60..e3b3dad6b0c7d59132b3cfaafe2d32abf6b1f5bb)*

samueltardieu commented on clippy_lints/src/operators/decimal_bitwise_operands.rs:
```diff
@@ -0,0 +1,77 @@
+            "use binary ({}), hex ({}), or octal ({}) notation for better readability",
+            numeric_literal::format(&format!("0b{val:b}"), None, false),
+            numeric_literal::format(&format!("0x{val:x}"), None, false),
+            numeric_literal::format(&format!("0o{val:o}"), None, false),
```
Rust knows how to format bin/oct/hex with a prefix:

```suggestion
            numeric_literal::format(&format!("{val:#b}"), None, false),
            numeric_literal::format(&format!("{val:#x}"), None, false),
            numeric_literal::format(&format!("{val:#o}"), None, false),
```

Artur-Sulej commented:
@llogiq Sorry to hear that, good luck with everything!

@llogiq @ada4a Commits squashed and rebased. Thanks for the review!

samueltardieu reviewed (commented):
It would be interesting to preserve the suffix (for example `254_u32` should suggest `0xfe_u32`, etc.).

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15215/3efb2acee79b25d664768e9a3566aea78623231b..273deea37c94fc1854f0c9e4d2552bc875affee6)*

Artur-Sulej commented on tests/ui/decimal_bitwise_operands.stderr:
```diff
@@ -0,0 +1,204 @@
+LL |     0b1010 & 99;
+   |              ^^
+   |
+   = help: use binary (0b110_0011), hex (0x0063), or octal (0o143) notation for better readability
```
Technically this is because the lint uses the shared `numeric_literal` helper, which is used throughout Clippy to format literals in a consistent way. For hex it groups digits by 4 and adds zero padding.

My understanding is that this was done for readability. Each hex digit represents 4 bits, so 4 digits are 16 bits, which matches common integer sizes (u16, i32, u64 etc). 
Padding to 4 digits (2 bytes) also avoids a dangling half-byte when you think in terms of 16‑bit.

To sum up, the benefits are: consistency (same formatting as elsewhere) and readability (visual grouping of bytes).

Artur-Sulej replied:
I did some archeology and found when it was introduced:
* https://github.com/rust-lang/rust-clippy/issues/2300
* https://github.com/rust-lang/rust-clippy/pull/2813

pitaj replied:
That's fine for now. I might come back later and change it to handle 1-2 digits better. No reason to delay this though.

pitaj replied:
That's fine for now. I might come back later and change it to handle 1-2 digits better. No reason to delay this though.

Artur-Sulej commented on clippy_lints/src/operators/mod.rs:
````diff
@@ -935,6 +936,28 @@ declare_clippy_lint! {
+    /// ```
+    #[clippy::version = "1.93.0"]
+    pub DECIMAL_BITWISE_OPERANDS,
+    nursery,
````
Done and rebased ✅

Artur-Sulej commented:
🔥
